Create rule S7027: Dependency cycles between classes should be removed (#4077)
SonarSource/rspec: rules/S7027/java/rule.adoc

== Why is this an issue?

The presence of cyclic dependencies between classes is detrimental to software quality attributes such as understandability, testability, 
reusability, and maintainability.

> Arguably, in a cycle of classes, there is no reasonable starting point, and none of the classes involved makes sense on its own.
> 

Reviewer: I would not mention that here and remove it without any replacement. This is an argument why it is hard to find the correct class in the cycle to report on, but not why we should avoid cycles. In the domain of the software, the classes of the data model can make sense even when they from a cycle (such as in the code example below: "Order", "Customer" and "Product")

```suggestion
```

=== What is the potential impact?

> Having these cycles greatly raises the coupling between components of the codebase, making it extremely hard to have separate testing, reuse or release.

Reviewer: We should avoid superlatives (greatly, extremely). Also, adding recursion as a reason:

```suggestion
Cyclic dependencies increase the coupling between the components of the codebase, making the software's architecture harder to understand and extend.

Also, they increase the likelihood of code to run into recursion problems.
```

Furthermore, if we try to visualize our codebase as a graph, to understand its structure, having cyclic dependencies could make the task nearly impossible.
On the other hand, the absence of them would result in a tree structure, where every component relies only on underlying components, and understanding the flow of the program will be much easier.

== How to fix it

=== Code examples

==== Noncompliant code example

> [source,java,diff-id=1,diff-type=noncompliant]
> ----
> FIXME
> ----
> 
> ==== Compliant solution
> 
> [source,java,diff-id=1,diff-type=compliant]
> ----
> FIXME
> ----

Reviewer: Suggestion for an example:
```suggestion
The following code contains two cycles: Order &rarr; Customer &rarr; Order, and Order &rarr; Product &rarr; Order. Furthermore, both cycles are connected to each other via class Order. This is called _a tangle consisting of two cycles_.

[source,java,diff-id=1,diff-type=noncompliant]
----
class Order {
    public Customer curstomer
    public List<Product> products    
}

class Customer {
    public List<Order> orders
}

class Product {
    public List<Order> orders
}
----

==== Compliant solution

The issue can be resolved by refacturing the architecture. The dependencies Curstomer &rarr; Order and Product &rarr; Order have been replaced by two service functions.

[source,java,diff-id=1,diff-type=compliant]
----
class Order {
  public Customer curstomer;
  public List<Order> products;
}

class Customer { }

class Product { }

interface Service {
  List<Order> getOrdersByCustomer(Customer customer);
  List<Order> getOrdersByProduct(Product product);
}
----
```
